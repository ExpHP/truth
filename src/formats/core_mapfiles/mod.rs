use ::std::collections::BTreeMap;

use crate::game::Game;
use crate::eclmap::Eclmap;

pub mod anm;
pub mod msg;
pub mod std;

/// Struct for representing some embedded mapfile information (notably the default signatures and
/// register types for all mainstream games).
///
/// Basically, the current mapfile syntax has a bit of a problem that it is impossible to factor
/// out into reusable parts.  And since some formats frequently change signatures of individual
/// instructions, putting signatures there requires maintaining an unduly amount of copy-pasta.
///
/// This internal data structure is designed to be easier to refactor and maintain compared to
/// an actual mapfile, and in the future may even help guide attempts to solve the factoring
/// problem for user mapfile syntax.
pub struct CoreSignatures {
    /// A simple method of factoring out a single [`CoreSignatures`] into reusable
    /// portions when other methods fail to suffice. There's no specific, prescribed usage.
    ///
    /// Each item listed here will be applied in sequence before the definitions contained on
    /// the struct itself. (with later values overriding earlier ones)
    ///
    /// Note that a significant amount of boilerplate can be avoided by making use of the
    /// "minimum game" fields on [`Self::ins`] and [`Self::var`], and so some formats don't
    /// actually need to make use of this at all.
    inherit: &'static [&'static CoreSignatures],

    /// A set of tuples, each the addition of a new instruction or the change of an existing
    /// instruction's signature in some game.  (a `None` indicates removal; i.e. the jumptable
    /// entry now points to the default branch)
    ///
    /// The strings are mapfile string signatures.
    ///
    /// When converted to a mapfile, the program will run down this list and apply each item
    /// from the current game or earlier.  The presence of these "minimum game" fields allows
    /// a single [`CoreSignatures`] to be easily applied to an entire range of games.
    ins: &'static [(Game, u16, Option<&'static str>)],

    /// Like [`Self::ins`] but for registers.
    var: &'static [(Game, i32, Option<&'static str>)],
}

impl CoreSignatures {
    pub fn to_mapfile(&self, game: Game) -> Eclmap {
        let mut mapfile = Eclmap::new();
        self.apply_to_mapfile(game, &mut mapfile);
        mapfile
    }

    fn apply_to_mapfile(&self, game: Game, mapfile: &mut Eclmap) {
        for parent in self.inherit {
            parent.apply_to_mapfile(game, mapfile);
        }

        for &(min_game, opcode, siggy_str) in self.ins {
            if min_game <= game {
                insert_or_remove(&mut mapfile.ins_signatures, opcode as _, siggy_str);
            }
        }

        for &(min_game, reg_id, type_str) in self.var {
            if min_game <= game {
                insert_or_remove(&mut mapfile.gvar_types, reg_id as _, type_str);
            }
        }
    }
}

fn insert_or_remove<K, V>(map: &mut BTreeMap<K, V::Owned>, key: K, value: Option<&V>)
where
    K: Ord + Eq,
    V: ToOwned + ?Sized,
{
    if let Some(value) = value {
        map.insert(key, value.to_owned());
    } else {
        map.remove(&key);
    }
}
