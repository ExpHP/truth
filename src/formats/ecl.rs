use std::fmt;
use std::io;
use std::num::NonZeroU64;

use enum_map::EnumMap;
use indexmap::{IndexSet, IndexMap};

use crate::ast;
use crate::io::{BinRead, BinWrite, BinReader, BinWriter, Encoded, ReadResult, WriteResult, DEFAULT_ENCODING};
use crate::diagnostic::{Emitter};
use crate::error::{GatherErrorIteratorExt, ErrorReported};
use crate::game::Game;
use crate::ident::{Ident, ResIdent};
use crate::llir::{self, ReadInstr, RawInstr, InstrFormat, IntrinsicInstrKind};
use crate::meta::{self, FromMeta, FromMetaError, Meta, ToMeta};
use crate::pos::{Sp, Span};
use crate::value::{ScalarType};
use crate::context::CompilerContext;
use crate::passes::DecompileKind;
use crate::resolve::RegId;

// =============================================================================

/// Game-independent representation of an ANM file.
#[derive(Debug, Clone)]
pub struct OldeEclFile {
    pub subs: IndexMap<Ident, Vec<RawInstr>>,
    pub timelines: IndexMap<Ident, Vec<RawInstr>>,
    /// Filename of a read binary file, for display purposes only.
    binary_filename: Option<String>,
}

impl OldeEclFile {
    // pub fn decompile_to_ast(&self, game: Game, ctx: &mut CompilerContext, decompile_kind: DecompileKind) -> Result<ast::ScriptFile, ErrorReported> {
    //     let emitter = ctx.emitter.while_decompiling(self.binary_filename.as_deref());
    //     decompile(self, &emitter, &game_format(game), ctx, decompile_kind)
    // }
    //
    // pub fn compile_from_ast(game: Game, ast: &ast::ScriptFile, ctx: &mut CompilerContext) -> Result<Self, ErrorReported> {
    //     compile(&game_format(game), ast, ctx)
    // }

    pub fn write_to_stream(&self, w: &mut BinWriter, game: Game) -> WriteResult {
        let emitter = w.emitter();
        write_olde_ecl(w, &emitter, &game_format(game)?, self)
    }

    pub fn read_from_stream(r: &mut BinReader, game: Game) -> ReadResult<Self> {
        let emitter = r.emitter();
        read_olde_ecl(r, &emitter, &game_format(game)?)
    }
}

// =============================================================================

#[cfg(nope)]
fn decompile(
    anm_file: &OldeEclFile,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
    ctx: &mut CompilerContext,
    decompile_kind: DecompileKind,
) -> Result<ast::ScriptFile, ErrorReported> {

}

// =============================================================================

#[cfg(nope)]
fn compile(
    format: &OldeFileFormat,
    ast: &ast::ScriptFile,
    ctx: &mut CompilerContext,
) -> Result<OldeEclFile, ErrorReported> {

}

// =============================================================================

fn read_olde_ecl(
    reader: &mut BinReader,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
) -> ReadResult<OldeEclFile> {
    let instr_format = &*format.instr_format();
    let timeline_format = &*format.timeline_format();

    let start_pos = reader.pos()?;

    if let Some(magic) = format.magic() {
        reader.expect_magic(emitter, &magic.to_le_bytes())?;
    }

    let num_subs = reader.read_u16()? as usize;
    let num_subs_high_word = reader.read_u16()? as usize;

    let timeline_array_len; // how many u32s to read for timelines
    let expected_nonzero_timelines; // how many of them SHOULD be nonzero
    match format.timeline_array_kind() {
        TimelineArrayKind::Pofv => {
            timeline_array_len = num_subs_high_word;
            expected_nonzero_timelines = num_subs_high_word;
        },
        TimelineArrayKind::Pcb { cap } => {
            timeline_array_len = cap;
            expected_nonzero_timelines = num_subs_high_word + 1;
        }
        TimelineArrayKind::Eosd { cap } => {
            timeline_array_len = cap;
            expected_nonzero_timelines = 1;
            if num_subs_high_word != 0 {
                emitter.emit(warning!(
                    "unexpected nonzero high word for num_subs: {:#x}", num_subs_high_word,
                )).ignore();
            }
        },
    };
    let timeline_offsets = reader.read_u32s(timeline_array_len)?;
    let sub_offsets = reader.read_u32s(num_subs)?;

    // expect a prefix of the timeline array to be filled
    // (because we can't represent a sparsely filled timeline array in the AST)
    let mut num_timelines = timeline_offsets.iter().position(|&x| x == 0).unwrap_or(timeline_offsets.len());
    for &offset in &timeline_offsets[num_timelines..] {
        if offset != 0 {
            return Err(emitter.emit(error!("unexpected timeline offset {:#x} after a null offset", offset)));
        }
    }

    if num_timelines != expected_nonzero_timelines {
        emitter.emit(warning!(
            "expected {} nonzero entries in timeline table, but found {}",
            expected_nonzero_timelines, num_timelines,
        )).ignore();
    }
    if matches!(format.timeline_array_kind(), TimelineArrayKind::Pcb { .. }) {
        num_timelines -= 1;  // in these games, that last entry points to the end of the file
    }

    eprintln!("NUM_SUBS {}", num_subs);
    eprintln!("NUM_TIMELINES {}", num_timelines);

    let subs = sub_offsets.into_iter().enumerate().map(|(index, sub_offset)| {
        let name = auto_sub_name(index as u32);

        eprintln!(": SUB {} AT {:#x}", index, sub_offset);
        reader.seek_to(start_pos + sub_offset as u64)?;
        let instrs = emitter.chain_with(|f| write!(f, "in sub {}", index), |emitter| {
            eprintln!("    TIME _OP_ SIZE DIFF MASK  ARGS");
            llir::read_instrs(reader, emitter, instr_format, sub_offset as u64, None)
        })?;
        Ok((name, instrs))
    }).collect::<ReadResult<_>>()?;

    let timelines = timeline_offsets[..num_timelines].iter().enumerate().map(|(index, &sub_offset)| {
        let name = auto_timeline_name(index as u32);

        eprintln!(": TIMELINE {} AT {:#x}", index, sub_offset);
        reader.seek_to(start_pos + sub_offset as u64)?;
        let instrs = emitter.chain_with(|f| write!(f, "in timeline sub {}", index), |emitter| {
            eprintln!("TIME ARG0 _OP_ SZ DF  ARGS");
            llir::read_instrs(reader, emitter, timeline_format, sub_offset as u64, None)
        })?;
        Ok((name, instrs))
    }).collect::<ReadResult<_>>()?;

    let binary_filename = Some(reader.display_filename().to_string());
    Ok(OldeEclFile { subs, timelines, binary_filename })
}

pub fn auto_sub_name(i: u32) -> Ident {
    format!("sub{}", i).parse::<Ident>().unwrap()
}
pub fn auto_timeline_name(i: u32) -> Ident {
    format!("timeline{}", i).parse::<Ident>().unwrap()
}

fn write_olde_ecl(
    w: &mut BinWriter,
    emitter: &impl Emitter,
    format: &OldeFileFormat,
    ecl: &OldeEclFile,
) -> WriteResult {
    let instr_format = &*format.instr_format();
    let timeline_format = &*format.timeline_format();

    let start_pos = w.pos()?;

    if let Some(magic) = format.magic() {
        w.write_u32(magic)?;
    }

    if ecl.timelines.len() > format.timeline_array_kind().max_timelines() {
        // FIXME: NEEDSTEST for each game
        return Err(emitter.emit(error!("too many timelines! (max allowed in this game is {})", ecl.timelines.len())));
    }

    match format.timeline_array_kind() {
        | TimelineArrayKind::Pofv { .. }
        | TimelineArrayKind::Pcb { .. } => {
            w.write_u16(ecl.subs.len() as _)?;
            w.write_u16(ecl.timelines.len() as _)?;
        },
        | TimelineArrayKind::Eosd { .. } => {
            w.write_u16(ecl.subs.len() as _)?;
            w.write_u16(0)?;
        },
    };

    let timeline_array_len = match format.timeline_array_kind() {
        TimelineArrayKind::Pcb { cap } => cap,
        TimelineArrayKind::Eosd { cap, .. } => cap,
        TimelineArrayKind::Pofv => ecl.timelines.len(),
    };

    let script_offsets_pos = w.pos()?;
    for _ in 0..ecl.subs.len() + timeline_array_len {
        w.write_u32(0)?;
    }

    let mut sub_offsets = vec![];
    for (index, (ident, sub)) in ecl.subs.iter().enumerate() {
        sub_offsets.push(w.pos()? - start_pos);
        emitter.chain_with(|f| write!(f, "in sub {} (index {})", ident, index), |emitter| {
            llir::write_instrs(w, emitter, instr_format, &sub)
        })?;
    }

    let mut timeline_offsets = vec![];
    for (index, (ident, sub)) in ecl.timelines.iter().enumerate() {
        timeline_offsets.push(w.pos()? - start_pos);
        emitter.chain_with(|f| write!(f, "in sub {} (index {})", ident, index), |emitter| {
            llir::write_instrs(w, emitter, timeline_format, &sub)
        })?;
    }

    let end_pos = w.pos()?;
    if matches!(format.timeline_array_kind(), TimelineArrayKind::Pcb { .. }) {
        // in these games, that last entry points to the end of the file
        timeline_offsets.push(end_pos - start_pos);
    }
    timeline_offsets.resize(timeline_array_len, 0);

    w.seek_to(script_offsets_pos)?;
    for offset in timeline_offsets {
        w.write_u32(offset as u32)?;
    }
    for offset in sub_offsets {
        w.write_u32(offset as u32)?;
    }
    w.seek_to(end_pos)?;
    Ok(())
}

// =============================================================================

fn game_format(game: Game) -> Result<OldeFileFormat, ErrorReported> {
    match game {
        | Game::Th06 | Game::Th07 | Game::Th08 | Game::Th09 | Game::Th095
        => Ok(OldeFileFormat { game }),

        _ => Err(unimplemented!("game {} not yet supported", game)),
    }
}

pub fn game_core_mapfile(game: Game) -> crate::Eclmap {
    super::core_mapfiles::msg::core_signatures(game).to_mapfile(game)
}

// =============================================================================

struct OldeFileFormat { game: Game }

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
enum TimelineArrayKind {
    /// There appears to be space for `cap` items, but only the first is used.
    Eosd { cap: usize },
    /// The timeline offset array is a fixed size, but the file specifies how many are actually used.
    ///
    /// (there will be that many nonzero entries *plus one*; the final one will be the ending offset of
    ///  the last timeline, which should coincide with the size of the file)
    Pcb { cap: usize },
    /// The length of the timeline offset array is specified in the file.  All entries are used.
    Pofv,
}

impl TimelineArrayKind {
    fn max_timelines(&self) -> usize {
        match *self {
            TimelineArrayKind::Eosd { .. } => 1,
            TimelineArrayKind::Pcb { cap } => cap - 1, // room for file size
            TimelineArrayKind::Pofv => usize::MAX,
        }
    }
}

impl OldeFileFormat {
    fn magic(&self) -> Option<u32> {
        match self.game {
            | Game::Th06 | Game::Th07 => None,
            | Game::Th08 | Game::Th095 => Some(0x00_00_08_00),
            | Game::Th09 => Some(0x00_00_09_00),
            _ => unimplemented!("game not yet supported"),
        }
    }

    fn timeline_array_kind(&self) -> TimelineArrayKind {
        match self.game {
            | Game::Th06 => TimelineArrayKind::Eosd { cap: 3 },

            | Game::Th07 | Game::Th08 | Game::Th095
            => TimelineArrayKind::Pcb { cap: 16 },

            | Game::Th09 => TimelineArrayKind::Pofv,

            _ => unimplemented!("game not yet supported"),
        }
    }

    fn instr_format(&self) -> Box<dyn InstrFormat> { Box::new(InstrFormat06) }
    fn timeline_format(&self) -> Box<dyn InstrFormat> { Box::new(TimelineInstrFormat { game: self.game }) }
}

struct InstrFormat06;

impl InstrFormat for InstrFormat06 {
    fn intrinsic_opcode_pairs(&self) -> Vec<(llir::IntrinsicInstrKind, u16)> {
        vec![]
    }

    fn instr_header_size(&self) -> usize { 12 }

    fn read_instr(&self, f: &mut BinReader, _: &dyn Emitter) -> ReadResult<ReadInstr> {
        let time = f.read_i32()?;
        let opcode = f.read_u16()?;
        let size = f.read_u16()? as usize;
        let difficulty = f.read_u16()?;
        let param_mask = f.read_u16()?;

        let args_blob = f.read_byte_vec(size - self.instr_header_size())?;

        eprintln!("{:08x} {:04x} {:04x} {:04x} {:04x}  {}", time, opcode, size, difficulty, param_mask, crate::io::hexify(&args_blob));
        let instr = RawInstr { time, opcode, param_mask, args_blob, difficulty, ..RawInstr::DEFAULTS };

        if opcode == (-1_i16) as u16 {
            Ok(ReadInstr::Terminal)
        } else {
            Ok(ReadInstr::Instr(instr))
        }
    }

    fn write_instr(&self, f: &mut BinWriter, _: &dyn Emitter, instr: &RawInstr) -> WriteResult {
        f.write_i32(instr.time)?;
        f.write_u16(instr.opcode)?;
        f.write_u16(self.instr_size(instr) as _)?;
        f.write_u16(instr.difficulty)?;
        f.write_u16(instr.param_mask)?;
        f.write_all(&instr.args_blob)?;
        Ok(())
    }

    fn write_terminal_instr(&self, f: &mut BinWriter, _: &dyn Emitter) -> WriteResult {
        f.write_i32(-1)?; // time
        f.write_i16(-1)?; // opcode
        f.write_u16(self.instr_header_size() as _)?; // size
        f.write_u16(0xff00)?; // difficulty
        f.write_u16(0x00ff)?; // param_mask
        Ok(())
    }
}

struct TimelineInstrFormat { game: Game }
impl TimelineInstrFormat {
    /// In some games, the terminal instruction is shorter than an actual instruction.
    fn has_short_terminal(&self) -> bool {
        self.game < Game::Th08
    }
}

impl InstrFormat for TimelineInstrFormat {
    fn intrinsic_opcode_pairs(&self) -> Vec<(llir::IntrinsicInstrKind, u16)> {
        vec![]
    }

    fn instr_header_size(&self) -> usize { 8 }

    fn read_instr(&self, f: &mut BinReader, emitter: &dyn Emitter) -> ReadResult<ReadInstr> {
        let time = f.read_i16()? as i32;
        let arg_0 = f.read_i16()? as i32;

        // with some games the terminal instruction is only 4 bytes long so we must check now
        if self.has_short_terminal() && (time, arg_0) == (-1, 4) {
            // FIXME: really should be MaybeTerminal since both arg_0 = 4 and time = -1
            //        are things that naturally occur on real instructions.
            return Ok(ReadInstr::Terminal);
        }

        let opcode = f.read_u16()?;
        let size = f.read_u8()? as usize;
        let difficulty = f.read_u8()? as u16;

        // in games with the normal-sized terminal, the size is incorrectly 0, so check before reading args
        if !self.has_short_terminal() && (time, arg_0, opcode, size, difficulty) == (-1, -1, 0, 0, 0) {
            // FIXME: really should be MaybeTerminal
            return Ok(ReadInstr::Terminal);
        }

        let args_blob = f.read_byte_vec(size - self.instr_header_size())?;
        eprintln!("{:04x} {:04x} {:04x} {:02x} {:02x}  {}", time as i16, arg_0 as u16, opcode as u16, size as u8, difficulty as u8, crate::io::hexify(&args_blob));

        let instr = RawInstr {
            time, opcode, difficulty, args_blob,
            extra_arg: Some(arg_0 as _),
            ..RawInstr::DEFAULTS
        };
        Ok(ReadInstr::Instr(instr))
    }

    fn write_instr(&self, f: &mut BinWriter, _: &dyn Emitter, instr: &RawInstr) -> WriteResult {
        f.write_i16(instr.time as _)?;
        f.write_i16(instr.extra_arg.expect("missing extra_arg on timeline?!") as _)?;
        f.write_u16(instr.opcode)?;
        f.write_u8(self.instr_size(instr) as _)?;
        f.write_u8(instr.difficulty as _)?;
        f.write_all(&instr.args_blob)?;
        Ok(())
    }

    fn write_terminal_instr(&self, f: &mut BinWriter, _: &dyn Emitter) -> WriteResult {
        if self.game < Game::Th08 {
            f.write_i16(-1)?; // time
            f.write_i16(4)?; // arg_0
        } else {
            f.write_i16(-1)?; // time
            f.write_i16(-1)?; // arg_0
            f.write_u32(0)?; // opcode, size, difficulty
        }
        Ok(())
    }
}
