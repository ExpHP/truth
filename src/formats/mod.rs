pub mod anm;
pub mod std;
pub mod msg;
pub mod mission;
pub mod ecl;

mod core_mapfiles;
