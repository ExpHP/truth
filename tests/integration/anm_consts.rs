use crate::integration_impl::formats::*;
use crate::integration_impl::{TestFile, expected};

// This is the first written test of const vars that depend on other const vars.
// (sprite IDs were the first const vars implemented in the compiler, even before
//  const var items!)
source_test!(
    ANM_12, sprite_ids_gone_wild,
    full_source: r#"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {
        valueA: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: valueC + 2 - valueC},
        valueB: {x: 0.0, y: 0.0, w: 0.0, h: 0.0},
        valueC: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 26 * 2 + 1},
        valueD: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: valueE - 1},
    },
}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {
        valueE: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 401},
        valueF: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: _S(%valueE + 2.4) + 1},
    },
}

script script0 {
    ins_1();
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].sprites[0].id, Some(2));
        assert_eq!(anm.entries[0].sprites[1].id, None);  // 3
        assert_eq!(anm.entries[0].sprites[2].id, Some(53));
        assert_eq!(anm.entries[0].sprites[3].id, Some(400));
        assert_eq!(anm.entries[1].sprites[0].id, None);  // 401
        assert_eq!(anm.entries[1].sprites[1].id, Some(404));
    },
);

source_test!(
    ANM_10, meta_sprite_id_circular_dependency,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: bestSprite * 3},
        bestSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: coolestSprite * 3},
        coolestSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: coolSprite + 1},
    },
}
    "#,
    expect_error: "depends on its own value",
);

source_test!(
    ANM_10, meta_non_const,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    memory_priority: 3 * I0,
    low_res_scale: false,
    sprites: {
        sprite200: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 200},
    },
}
    "#,
    // NOTE: be careful changing this.  If the new error complains about missing fields
    // or missing image data, fix the test input instead.
    expect_error: "const",
);

// It is okay for two sprites to have the same name (this occurs in decompiled output),
// but they must also have the same ID.
source_test!(
    ANM_12, sprite_ids_dupe,
    full_source: r#"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {
        valueA: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 26 * 2 + 1},
    },
}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {
        xyzzyx: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 24},
        valueA: {x: 1.0, y: 1.0, w: 1.0, h: 1.0, id: 53},
    },
}

script script0 {
    ins_1();
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].sprites[0].id, Some(53));
        assert_eq!(anm.entries[1].sprites[0].id, Some(24));
        assert_eq!(anm.entries[1].sprites[1].id, Some(53));
    },
);

// same but one's implicit
source_test!(
    ANM_12, sprite_ids_dupe_implicit,
    full_source: r#"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {
        valueA: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 26 * 2 + 1},
    },
}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {
        xyzzyx: {x: 0.0, y: 0.0, w: 0.0, h: 0.0, id: 52},
        valueA: {x: 1.0, y: 1.0, w: 1.0, h: 1.0},   // dupe, but has same id as above
    },
}

script script0 {
    ins_1();
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].sprites[0].id, Some(53));
        assert_eq!(anm.entries[1].sprites[0].id, Some(52));
        assert_eq!(anm.entries[1].sprites[1].id, None);  // 53
    },
);

// Now they have mismatched IDs.
source_test!(
    ANM_12, err_sprite_clash,
    full_source: r#"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {my_sprite: {x: 1.0, y: 1.0, w: 111.0, h: 111.0, id: 0}},
}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {my_sprite: {x: 2.0, y: 2.0, w: 222.0, h: 220.0, id: 1}},
}

script script0 {
    ins_1();
}
    "#,
    expect_error: "ambiguous"
);

// Same but one's implicit.
source_test!(
    ANM_12, err_sprite_clash_implicit,
    full_source: r#"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {my_sprite: {x: 1.0, y: 1.0, w: 111.0, h: 111.0, id: 1}},
}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {my_sprite: {x: 2.0, y: 2.0, w: 222.0, h: 220.0}},
}

script script0 {
    ins_1();
}
    "#,
    expect_error: "ambiguous"
);

// Type-checking/const-checking sprite IDs is actually a bit annoying.
source_test!(
    ANM_10, meta_sprite_id_type_error,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3.5},
    },
}
    "#,
    expect_error: expected::TYPE_ERROR,
);

source_test!(
    ANM_10, meta_sprite_id_non_const,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    sprites: {
        sprite200: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3 * I0},
    },
}
    "#,
    expect_error: "const",
);

// Dupes may follow different code paths for type checking and const checking since
// the expressions are never used beyond checking equality to the originals.
source_test!(
    ANM_10, meta_sprite_id_dupe_type_error,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3},
    },
}
entry {
    path: "subdir/file-3.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3.5},
    },
}
    "#,
    expect_error: expected::TYPE_ERROR,
);

source_test!(
    ANM_10, meta_sprite_id_dupe_non_const,
    items: r#"
entry {
    path: "subdir/file-2.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3},
    },
}
entry {
    path: "subdir/file-3.png",
    has_data: false,
    sprites: {
        coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 3 * I0},
    },
}
    "#,
    expect_error: "const",
);

source_test!(
    ANM_12, const_using_sprite_id,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

entry {
    path: "lmao.png",
    has_data: false,
    sprites: {
        sprite0: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 10},
    },
}

const int B = sprite0;

script script0 {
    ins_3(B);
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].scripts[0].instrs[0].args_blob, vec![10, 0, 0, 0]);
    },
);

source_test!(
    ANM_12, sprite_id_using_const,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

entry {
    path: "lmao.png",
    has_data: false,
    sprites: {
        sprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: B},
    },
}

const int B = 10;

script script0 {
    ins_3(sprite);
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].scripts[0].instrs[0].args_blob, vec![10, 0, 0, 0]);
    },
);

source_test!(
    ANM_12, const_shadows_sprite_id,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

entry {
    path: "lmao.png",
    has_data: false,
    sprites: {
        B: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 42},
        C: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: B + 2},
    },
}

const int B = 10;

script script0 {
    ins_3(B);
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].sprites[0].id, Some(42));
        assert_eq!(anm.entries[0].sprites[1].id, Some(12));
        assert_eq!(anm.entries[0].scripts[0].instrs[0].args_blob, vec![10, 0, 0, 0]);
    },
);

source_test!(
    ANM_12, consts_in_various_other_meta,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

const string FILEPATH = "lmao.png";

entry {
    path: FILEPATH,
    has_data: false,
    sprites: {
        sprite0: {x: POS_X, y: 0.0, w: 512.0, h: 480.0},
        sprite1: {x: POS_X + 3.0, y: 0.0, w: 512.0, h: 480.0},
    },
}

const float POS_X = 20.0;

script script0 {}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].sprites[0].offset[0], 20.0);
        assert_eq!(anm.entries[0].sprites[1].offset[0], 23.0);
    },
);

source_test!(
    ANM_12, sprite_script_name_clash,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

entry {
    path: "lmao.png",
    has_data: false,
    sprites: {
        wild: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 22 * 2 - 2},
    },
}

script 23 script23 {
    ins_3(-1);
}

script wild {  // should have ID 1
    ins_3(-1);
}
    "#,
    expect_error: "ambiguous",
);

source_test!(
    ANM_12, sprite_shadows_reg_alias,
    full_source: r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "tests/integration/resources/th12-embedded-image-source.anm"

entry {
    path: "lmao.png",
    has_data: false,
    sprites: {
        RAND: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 22 * 2 - 2},
    },
}

const int x = $RAND;

script 23 script23 {
    ins_3(x);
}
    "#,
    check_compiled: |output, format| {
        let anm = output.read_anm(format);
        assert_eq!(anm.entries[0].scripts[0].instrs[0].args_blob, vec![42, 0, 0, 0]);
    },
);

const SCRIPT_IDS_EXAMPLE: &'static str = r#"
#pragma mapfile "map/any.anmm"
#pragma image_source "./tests/integration/resources/th12-multiple-match-source.anm"

entry {
    path: "subdir/file1.png",
    has_data: false,
    sprites: {
        sprite0: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 0},
    },
}

script myScript {
    scriptNew(child);
}

script 24 irrelevant {}

entry {
    path: "subdir/file2.png",
    has_data: false,
    sprites: {
        sprite1: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: 1},
    },
}

script child {}

script 101 another {}
"#;

#[test]
fn script_ids() {
    let format = ANM_12;
    let source = TestFile::from_content("input", SCRIPT_IDS_EXAMPLE);
    let anm = format.compile(&source).read_anm(&format);

    assert_eq!(anm.entries[0].scripts[0].id, 0);
    assert_eq!(anm.entries[0].scripts[1].id, 24);
    assert_eq!(anm.entries[1].scripts[0].id, 25);
    assert_eq!(anm.entries[1].scripts[1].id, 101);
}

#[test]
fn scripts_as_consts() {
    let format = ANM_12;
    let source = TestFile::from_content("input", SCRIPT_IDS_EXAMPLE);
    let anm = format.compile(&source).read_anm(&format);

    // the value of the const should be the *index across all entries* (2), not the ID (25)
    assert_eq!(anm.entries[0].scripts[0].instrs[0].args_blob, vec![2, 0, 0, 0]);
}
