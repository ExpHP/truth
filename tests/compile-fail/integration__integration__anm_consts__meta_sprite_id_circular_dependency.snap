---
source: tests/integration/anm_consts.rs
expression: stderr
---
error: cycle in const definition
   ┌─ <input>:30:9
   │
30 │         coolSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: bestSprite * 3},
   │         ^^^^^^^^^^ cyclic const
   ·
33 │         coolestSprite: {x: 0.0, y: 0.0, w: 512.0, h: 480.0, id: coolSprite + 1},
   │                                                                 ---------- depends on its own value here


